python/revng/ptml/wrappers.py

```diff
@@ -0,0 +1,129 @@
+#
+# This file is distributed under the MIT License. See LICENSE.md for details.
+#
+
+import mmap
+import re
+import tarfile
+from contextlib import contextmanager, suppress
+from io import BytesIO, StringIO
+from typing import IO, Dict, Generator, List, Optional, Protocol, Tuple, Union, cast
+
+import yaml
+
+# Get yaml.CSafeLoader if present, fall back to yaml.SafeLoader otherwise
+YAMLLoader = getattr(yaml, "CSafeLoader", yaml.SafeLoader)
+
+Buffer = Union[str, bytes]
+FileLike = Union[IO[bytes], IO[str]]
+
+# When calling InputWrapper.keys the return is either
+# * A tuple with a lone `None`, indicating a single PTML document
+# * A list of string keys, for a multi-document container
+WrapperKeys = Union[Tuple[None], List[str]]
+
+
+class InputWrapper(Protocol):
```

Comment:
1. The "wrapper" terminology is a bit hard to grasp for me.
   This seems more like a `DocumentList`, `TarDocumentList`, `YAMLDocumentList` and `SingleDocumentList`.
2. These classes are not PTML-specific.
   Can we move them somewhere else more appropriate?

Reply:
1. I'll give it a think, but I don't have much against `*DocumentList`
2. There's the `_is_plain_ptml` which is required for things to work

Reply:
Yes, but then it's not as general as it could be, what I mean is that we can have:
1. A set of classes that wrap X, Dict[str, X] in yaml and Dict[str, X] in tar
2. A set of classes from (1) where `X = xml`
3. A set of classes from (1) where `X = PTML`

If we want some truly generic I'd avoid (2) because it basically has no benefit to (3).
If anything, I can pass a function like "is_single_element" to `wrap` so that at least it's a bit reusable